### PR TITLE
Query: Adds distributed query gateway client pipeline stage

### PR DESCRIPTION
**Pull request by neildsh**
## Description

This changes the `DistributedQueryPipelineStage` which acts as a client pipeline stage for a Distributed Query gateway. Customers can utilize this pipeline by setting an environment variable called `AZURE_COSMOS_DISTRIBUTED_QUERY_GATEWAY_ENABLED` to `True`. This pipeline stage uses an `ICosmosDistributedQueryClient` to talk to the distributed query gateway. For its part, this client just uses the `DocumentClient` to craft the requests using gateway mode. Integration tests are added in `DistributedQueryClientTests`. Since the distributed query gateway is not currently bundled as a part of the emulator, the only queries that can be integration tested are passthrough queries that are currently supported by the routing gateway.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Pagination/NetworkAttachedDocumentContainer.cs [hunk 1]
```diff
@@ -39,15 +40,17 @@ internal sealed class NetworkAttachedDocumentContainer : IMonadicDocumentContain
 
         public NetworkAttachedDocumentContainer(
             ContainerInternal container,
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Pagination/NetworkAttachedDocumentContainer.cs:42]: Please add another overload that does not take the distributedQueryClient

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 2]
```diff
@@ -187,10 +187,12 @@ public class QueryRequestOptions : RequestOptions
...
         internal TestInjections TestSettings { get; set; }
 
-        internal FeedRange FeedRange { get; set; }
-
+        internal FeedRange FeedRange { get; set; }
+
         internal bool IsNonStreamingOrderByQueryFeatureDisabled { get; set; } = ConfigurationManager.IsNonStreamingOrderByQueryFeatureDisabled(defaultValue: false);
 
+        internal bool EnableDistributedQueryGatewayMode { get; set; } = ConfigurationManager.IsDistributedQueryGatewayModeEnabled(defaultValue: false);
+
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs:195]: Please add a comment that this env variable is temporary

File: Microsoft.Azure.Cosmos/src/Query/Core/QueryClient/CosmosDistributedQueryClient.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    using Microsoft.Azure.Cosmos.Query.Core.Pipeline.Pagination;
+    using Microsoft.Azure.Cosmos.Resource.CosmosExceptions;
+    using Microsoft.Azure.Cosmos.Tracing.TraceData;
+    using Microsoft.Azure.Documents;
+
+    internal sealed class CosmosDistributedQueryClient : ICosmosDistributedQueryClient
+    {
+        private readonly CosmosClientContext cosmosClientContext;
+
+        private readonly string resourceLink;
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/QueryClient/CosmosDistributedQueryClient.cs:23]: >resourceLink [](http://example.com/codeflow?start=32&length=12)

containerLink

File: Microsoft.Azure.Cosmos/src/Query/Core/QueryClient/CosmosDistributedQueryClient.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+        private readonly Guid correlatedActivityId;
+
+        private readonly IDocumentClientRetryPolicy retryPolicy;
+
+        private DocumentClient DocumentClient => this.cosmosClientContext.DocumentClient;
+
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/QueryClient/CosmosDistributedQueryClient.cs:30]: Please remove this field

File: Microsoft.Azure.Cosmos/src/Query/Core/QueryClient/CosmosDistributedQueryClient.cs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    internal sealed class CosmosDistributedQueryClient : ICosmosDistributedQueryClient
+    {
+        private readonly CosmosClientContext cosmosClientContext;
+
+        private readonly string resourceLink;
+
+        private readonly Guid correlatedActivityId;
+
+        private readonly IDocumentClientRetryPolicy retryPolicy;
+
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/QueryClient/CosmosDistributedQueryClient.cs:28]: Make this a property

File: Microsoft.Azure.Cosmos/src/Pagination/NetworkAttachedDocumentContainer.cs (hunk 1 above)
Comment (neildsh) [reply Microsoft.Azure.Cosmos/src/Pagination/NetworkAttachedDocumentContainer.cs:42]: fixed

**Review by sboshra — APPROVED** [review]

**Review by Maya-Painter — APPROVED** [review]

**Pull request merged**